README.md
ByteMail
========

Peer-to-peer Email Protocol

Info
----

If you would like to help out the ByteMail network, open port 5333 and set the relay config to True in bytemail.py

Contact
-------

If you want to get in touch with me you can contact me over ByteMail, my address is: 9580a97489dd46b59b8119c210916c55

Website
-------

http://bytemailproject.org


CHANGELOG
Version 0.4.61

- Simple Fixes

Version 0.4.6

- Can no send a message to multiple people in the GUI seperating the addresses with a comma in the "To:" form. ex: Me, tom, ben
- Fixed replying bug.`

Version 0.4.5

- Updated Style
- Bug Fixes
- The nodes.db was getting cluttered so I decided to delete it and start over (this was due to a lot of testing) that network still exists but it will no longer add new nodes. If you want to join the new network you need to delete all your .db files (except for your addresses.db) and restart ByteMail

Version 0.4.3

- Added reply button to /read/
- Fixed Addressbook error

Version 0.4.2

- Added addressbook to WebUI
- Can no delete sent messages from webUI

Version 0.4.12

- Fixed /read/ layout error
- Added new emoticons - :frown:, :disapproval:, and :wink:

Version 0.4.11

- Fixed JSON Error in bytemailgui.py

Version 0.4.1

- Fixed new client bug
- GUI Fixes

Version 0.4.0 

- Added WebUI which starts by default
- Added KeyError checking now

Version 0.3.25 Beta

- Added sent time to sent.db database entries

Version 0.3.24 Beta

- Added emoticons! Current ones are: :flag: :smile: :radioactive: :pirate:

Version 0.3.23 Beta

- When a new message is received instead of displaying the address it will display the name you assigned to that address, assuming the address is in your addressbook (only applies to relay nodes)
- Added sent_purge command, deletes all entries in the sent database

Version 0.3.22 Beta

- Fixed an error with the check command when un-timestamed messages are in the database

Version 0.3.21 Beta

- Fixed multiple entries into the addressbook.db
- Added updating to addressbook (simple insert a new address with the same name to update it)
- Added remove_address command in order to delete from the addressbook

Version 0.3.2 Beta

- Added Timestamp to messages
- Added sent command for checking sent messages
- Added sent_delete command for deleting messages in the sent.db file

Version 0.3.1 Beta

- Fixed RSA signature bug
- Added version command

Version 0.3.00 Beta

- Added RSA Signatures to messages & message deleations

Version 0.2.89 Beta

- Blank entries can no longer be inserted into the addressbook.db
- Added new command "addresses" shows all addresses in your address book

Version 0.2.88 Beta

- Fixed Connection reset by peer error

Version 0.2.87 Beta

- Added purge command to delete all messages sent to the node

Version 0.2.86 Beta

- When enter is pressed the previous command is no longer executed
- Fixed invalid "To: " field bug

Version 0.2.85 Beta

- Fixed long message error bug
- Fixed checkin bug

Version 0.2.84 Beta

- Key size changed from 512 to 1024

Version 0.2.83 Beta

- Added documentation to commands

Version 0.2.82 Beta

- Added timeout to delete.py
- User is no longer notified when a node "checks in"
- General bug fixes

Version 0.2.81 Beta

- Spelling errors

Version 0.2.8 Beta

- RSA is now working side by side with AES so that there is no message limit
- RSA key shortened to 512 bytes for faster generation
- Better POW

Version 0.2.7 Beta

- Proof-of-Work has been added

Version 0.2.6 Beta

- eval() exploit now fixed.
- Bug fix in broker

Version 0.2.5 Beta * Major Update

- Encryption has now been added to messages! It's about time!
- Broker now transfers your public key as well
- On start a 4056 pubkey is created (may need to be edited)


Version 0.2.2 Beta

- Delete command fixed, and now working.
- No more collision of message IDs
- When a node is running outbound connections only the list of nodes is randomized to keep the load distributed

Version 0.2.1 Beta

- Delete command now deletes message on all nodes.

Version 0.2.0 Beta * Major Update

- Complete rewrite of system
- Now only outgoing connections are needed to connect to network.
- get_nodes and get_messages command added, used for sharing messages and nodes.
- Bug fixes, and other neat things

* In next update

  - Messages will be encrypted
  - Delete command which is sent to all nodes when a message is deleted  

Version 0.1.8

- ByteMail now checks for open ports again, this time it is done more efficiently.
- Now using SSL to transfer data.

Version 0.1.7

- Got rid of script that checks for open ports.

Version 0.1.6

- Startup now checking for open ports, if an open port is found ByteMail will run on that port. If not, ByteMail will default to port 5333
- General fixes

Version 0.1.47

- Bug Fix: If checkin addr is not 32 characters long the checkin fails.

Version 0.1.46

- Messages that are larger than 1Kb will now go through.

Version 0.1.45

- Bug Fixes
- Added exit command
- Ctrl+c no longer quits program
- Blank messages can no longer be sent

Version 0.1.4

- Now prevents multiple enteries of the same node into the database.

Version 0.1.3

- If the checkin command does not go through, it is added to the unsent database.
- Changed time between loops in unsent.py from 10 seconds to 60 seconds. 

Version 0.1.2

- Fixed database Bug
- Fixed formatting of read and check command
- Added delete command
- General clean up

Version 0.1.1

- If a node is not online the messages are stored in unsent.db and sent at a later time.
- There is now another thread started for unsent.py
- Every 10 seconds messages in unsent.db are attempted to be sent.

Version 0.1.0

- Initial release


bytemail.py
#!/usr/bin/python
import socket                                                                                                                         
import db
import checkin
import message
import read
import check
import json
import threading
import os
import uuid
import cmd
import thread
import unsent
import delete
import ssl
import get_messages
import get_nodes
import time
import random
import rsa
import addressbook
import sent
import bytemailgui
import webbrowser

__version__ = "0.4.6"
GUI = True

class ByteMail:
    
    def __init__(self, addr, pubkey):
        self.cmds = {
                "delete":delete.delete,
                "checkin":checkin.checkin,
                "message":message.message,
                "get_messages":get_messages.get_messages,
                "get_nodes":get_nodes.get_nodes,
                }
        self.broker = ("198.147.20.190", 4321)
        self.addr = addr
        self.pubkey = pubkey
        self.port = 5333
        self.host = "0.0.0.0"
        self.open_port = False
        self.config = {
                "relay":True
                }
    def main(self):
        if not db.nodes.find("nodes", "all"):
            check = self.config['relay']
            if check:
                self.open_port = True
                print "You are running as a relay node."
            db.data.insert("data", {"port":self.port})
            print "Downloading Nodes..."
            self.get_nodes()
            print "Checking in to nodes..."
            self.send_checkin()
            print "Done!"
        
        check = self.config['relay']
        if check:
            self.open_port = True
            print "You are running as a relay node."
        else:
            print "You are not running as a relay node."
        if self.open_port:
            sock = ssl.socket()
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            sock.bind((self.host, self.port))
            sock.listen(5)
            while True:
                obj, conn = sock.accept()
                threading.Thread(target=self.handle, args=(obj, conn[0])).start()
        else:
            while True:
                check = db.nodes.find("nodes", "all")
                random.shuffle(check)
                node = None
                for x in check:
                    s = ssl.socket()
                    try:
                        s.connect((x['ip'], x['port']))
                    except:
                        s.close()
                        continue
                    else:
                        node = (x['ip'], x['port'])
                        s.close()
                        break

                sock = ssl.socket()
                try:
                    sock.connect(node)
                except:
                    continue
                else:
                    while True:
                        try:
                            get_messages.send_get_messages(x['ip'], x['port'])
                            get_nodes.send_get_nodes(x['ip'], x['port'])
                            time.sleep(10)
                        except Exception, error:
                            sock.close()
                            break

    def handle(self, obj, ip):
        try:
            data = obj.recv(102400)
        except:
            obj.close()
            return
        if data:
            try:
                data = json.loads(data)
            except ValueError:
                obj.close()
            else:
                try:
                    self.cmds[data['cmd']](obj, ip, data)
                except KeyError:
                    pass
    def send_checkin(self):
        nodes = db.nodes.find("nodes", "all")
        for x in nodes:
            s = ssl.socket()
            try:
                s.settimeout(1)
                s.connect((x['ip'], x['port']))
                s.send(json.dumps({"cmd":"checkin", "addr":self.addr, "port":self.port, "publickey":self.pubkey}))
                s.close()
            except Exception, error:
                s.close()
                db.unsent.insert("unsent",  {"to":[x['ip'], x['port']], "message":{"cmd":"checkin", "addr":self.addr, "port":self.port, "publickey":self.pubkey}}) 

    def get_nodes(self):
        send = {"addr":self.addr, "port":self.port, "publickey":self.pubkey}
        s = ssl.socket()
        s.connect(self.broker)
        s.send(json.dumps(send))
        with open("nodes.db", 'wb') as file:
            while True:
                data = s.recv(1024)
                if data:
                    file.write(data)
                else:
                    break
        print "Downloaded Nodes!"

class Prompt(cmd.Cmd):
    prompt = "ByteMail$ " 
    def do_send(self, line):
        self.lastcmd = ""
        addr = db.data.find("data", "all")[0]['addr']
        to = raw_input("To: ")
        if to == "":
            return "You must fill out the field."
        if len(to) != 32:
            if len(str(addressbook.check_entry(to))) != 32:
                print "Address Invalid."
                return
            else:
                to = str(addressbook.check_entry(to))
        title = raw_input("Title: ")
        msg = raw_input("Message: ")
        if not to or not title or not msg:
            print "You need to fill out all the fields."
        else:
	    print message.send_msg(msg, title, to, addr)
    def do_version(self, line):
        print __version__
        self.lastcmd = ""
    def help_send(self):
        print "Sends a message."
        self.lastcmd = ""
    def do_check(self, line):
        addr = db.data.find("data", "all")[0]['addr']
        check_ = check.check(addr)
        if not check_:
            print "You have no messages."
        else:
            for x in check_:
                print x
        self.lastcmd = ""
    def help_check(self):
        print "Displays a list of messages"
        self.lastcmd = ""
    def do_read(self, id):
        addr = db.data.find("data", "all")[0]['addr']
        print read.read(id, addr).decode("utf-8")
        self.lastcmd = ""
    def help_read(self):
        print "Usage: read <id>"
        print "Reads the specified message ID"
        self.lastcmd = ""
    def do_addr(self, line):
        addr = db.data.find("data", "all")[0]['addr']
        print "Your address is:", addr
        self.lastcmd = ""
    def help_addr(self):
        print "Displays your ByteMail address"
        self.lastcmd = ""
    def do_add_address(self, line):
        self.lastcmd = ""	
        name = raw_input("Name: ")
        address = raw_input("Address: ")
        addressbook.add_entry(name,address)
    def help_addresses(self, line):
        self.lastcmd = ""
        print "Shows all addresses in your address book"
    def do_addresses(self, line):
        self.lastcmd = ""
        print addressbook.addresses()
    def help_add_address(self):
        print "Adds an address to your address book"

    def do_delete(self, line):
        self.lastcmd = ""
        addr = db.data.find("data", "all")[0]['addr']
        print delete.send_delete(line, addr)
    def help_delete(self):
        self.lastcmd = ""
        print "Usage: delete <id>"
        print "Deletes the message with the specified ID"

    def do_purge(self,line):
        print "Purging..."
        addr = db.data.find("data", "all")[0]['addr']
        print delete.send_delete_all(addr)
    def help_purge(self):
        print "Deletes all messages sent to this node."

    def do_exit(self, line):
        print "Bye!"
        exit()
    def help_exit(self):
        self.lastcmd = ""
        print "Closes ByteMail."
    def help_sent(self):
        self.lastcmd = ""
        print "Shows all sent messages."
    def do_sent(self, line):
        self.lastcmd = ""
        print sent.sent()
    def do_sent_delete(self, line):
        self.lastcmd = ""
        print sent.sent_delete(line)
    def help_remove_address(self):
        self.lastcmd = ""
        print "Delete an address from your addressbook. Usage: remove_address <entryname>"
    def do_remove_address(self, line):
        self.lastcmd = ""
        print addressbook.remove_address(line)
    def do_sent_purge(self, line):
        self.lastcmd = ""
        sent_ = db.sent.find("sent", "all")
        for x in sent_:
            sent.sent_delete(x['id'])
        print "Sent database purged."
if __name__ == "__main__":
    exists = db.data.find("data", "all")
    if not exists:
        print "First time running ByteMail"
        print "Generating new keys... This could take a while."
        publickey, privatekey = rsa.newkeys(1024)
        db.data.insert("data", {"addr":uuid.uuid4().hex, "publickey":str(publickey), "privatekey":str(privatekey)})
        db.messages.insert("messages", {})
        data = db.data.find("data", "all")[0]
        addr = data['addr']
        pubkey = data['publickey']
    else:
        data = db.data.find("data", "all")[0]
        addr = data['addr']
        pubkey = data['publickey']

    b = ByteMail(addr, pubkey)
    c = Prompt()
    thread.start_new_thread(b.main, ())
    thread.start_new_thread(unsent.unsent, ())
    if not GUI:
        while True:
            try:
                c.cmdloop()
            except KeyboardInterrupt:
                continue
    else:
        webbrowser.open("http://localhost:5334")
        bytemailgui.run()
